File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseTimeParser.cs
using System.Collections.Generic;
using System.Text.RegularExpressions;

using DateObject = System.DateTime;

namespace Microsoft.Recognizers.Text.DateTime
{
    public class BaseTimeParser : IDateTimeParser
    {
        public static readonly string ParserName = Constants.SYS_DATETIME_TIME; // "Time";

        private readonly ITimeParserConfiguration config;

        public BaseTimeParser(ITimeParserConfiguration configuration)
        {
            config = configuration;
        }

        public ParseResult Parse(ExtractResult result)
        {
            return this.Parse(result, DateObject.Now);
        }

        public DateTimeParseResult Parse(ExtractResult er, DateObject referenceTime)
        {
            object value = null;
            if (er.Type.Equals(ParserName))
            {
                DateTimeResolutionResult innerResult;

                // Resolve timezome
                if (TimeZoneUtility.ShouldResolveTimeZone(er, config.Options))
                {
                    var metadata = er.Data as Dictionary<string, object>;
                    var timezoneEr = metadata[Constants.SYS_DATETIME_TIMEZONE] as ExtractResult;
                    var timezonePr = config.TimeZoneParser.Parse(timezoneEr);

                    innerResult = InternalParse(er.Text.Substring(0, (int)(er.Text.Length - timezoneEr.Length)), referenceTime);

                    if (timezonePr != null && timezonePr.Value != null)
                    {
                        innerResult.TimeZoneResolution = ((DateTimeResolutionResult)timezonePr.Value).TimeZoneResolution;
                    }
                }
                else
                {
                    innerResult = InternalParse(er.Text, referenceTime);
                }

                if (innerResult.Success)
                {
                    innerResult.FutureResolution = new Dictionary<string, string>
                    {
                        { TimeTypeConstants.TIME, DateTimeFormatUtil.FormatTime((DateObject)innerResult.FutureValue) },
                    };

                    innerResult.PastResolution = new Dictionary<string, string>
                    {
                        { TimeTypeConstants.TIME, DateTimeFormatUtil.FormatTime((DateObject)innerResult.PastValue) },
                    };

                    value = innerResult;
                }
            }

            var ret = new DateTimeParseResult
            {
                Text = er.Text,
                Start = er.Start,
                Length = er.Length,
                Type = er.Type,
                Data = er.Data,
                Value = value,
                TimexStr = value == null ? string.Empty : ((DateTimeResolutionResult)value).Timex,
                ResolutionStr = string.Empty,
            };

            return ret;
        }

        public List<DateTimeParseResult> FilterResults(string query, List<DateTimeParseResult> candidateResults)
        {
            return candidateResults;
        }

        protected virtual DateTimeResolutionResult InternalParse(string text, DateObject referenceTime)
        {
            var innerResult = ParseBasicRegexMatch(text, referenceTime);
            return innerResult;
        }

        // parse basic patterns in TimeRegexList
        private DateTimeResolutionResult ParseBasicRegexMatch(string text, DateObject referenceTime)
        {
            var trimmedText = text.Trim().ToLowerInvariant();
            var offset = 0;

            var match = this.config.AtRegex.Match(trimmedText);

            if (!match.Success)
            {
                match = this.config.AtRegex.Match(this.config.TimeTokenPrefix + trimmedText);
                offset = this.config.TimeTokenPrefix.Length;
            }

            if (match.Success && match.Index == offset && match.Length == trimmedText.Length)
            {
                return Match2Time(match, referenceTime);
            }

            // parse hour pattern, like "twenty one", "16"
            // create a extract result which content the pass-in text
            int hour;
            if (this.config.Numbers.TryGetValue(text, out hour) || int.TryParse(text, out hour))
            {
                if (hour >= 0 && hour <= 24)
                {
                    var ret = new DateTimeResolutionResult();

                    if (hour == 24)
                    {
                        hour = 0;
                    }

                    if (hour <= Constants.HalfDayHourCount && hour != 0)
                    {
                        ret.Comment = Constants.Comment_AmPm;
                    }

                    ret.Timex = "T" + hour.ToString("D2");
                    ret.FutureValue = ret.PastValue =
                        DateObject.MinValue.SafeCreateFromValue(referenceTime.Year, referenceTime.Month, referenceTime.Day, hour, 0, 0);
                    ret.Success = true;
                    return ret;
                }
            }

            var regexes = this.config.TimeRegexes;

            foreach (var regex in regexes)
            {
                var exactMatch = regex.MatchExact(trimmedText, trim: true);

                if (exactMatch.Success)
                {
                    return Match2Time(exactMatch.Match, referenceTime);
                }
            }

            return new DateTimeResolutionResult();
        }

        private DateTimeResolutionResult Match2Time(Match match, DateObject referenceTime)
        {
            var ret = new DateTimeResolutionResult();
            bool hasMin = false, hasSec = false, hasAm = false, hasPm = false, hasMid = false;
            int hour = 0,
                min = 0,
                second = 0,
                day = referenceTime.Day,
                month = referenceTime.Month,
                year = referenceTime.Year;

            var writtenTimeStr = match.Groups["writtentime"].Value;

            if (!string.IsNullOrEmpty(writtenTimeStr))
            {
                // get hour
                var hourStr = match.Groups["hournum"].Value.ToLower();
                hour = this.config.Numbers[hourStr];

                // get minute
                var minStr = match.Groups["minnum"].Value;
                var tensStr = match.Groups["tens"].Value;

                if (!string.IsNullOrEmpty(minStr))
                {
                    min = this.config.Numbers[minStr];
                    if (!string.IsNullOrEmpty(tensStr))
                    {
                        min += this.config.Numbers[tensStr];
                    }

                    hasMin = true;
                }
            }
            else if (!string.IsNullOrEmpty(match.Groups["mid"].Value))
            {
                hasMid = true;
                if (!string.IsNullOrEmpty(match.Groups["midnight"].Value))
                {
                    hour = 0;
                    min = 0;
                    second = 0;
                }
                else if (!string.IsNullOrEmpty(match.Groups["midmorning"].Value))
                {
                    hour = 10;
                    min = 0;
                    second = 0;
                }
                else if (!string.IsNullOrEmpty(match.Groups["midafternoon"].Value))
                {
                    hour = 14;
                    min = 0;
                    second = 0;
                }
                else if (!string.IsNullOrEmpty(match.Groups["midday"].Value))
                {
                    hour = Constants.HalfDayHourCount;
                    min = 0;
                    second = 0;
                }
            }
            else
            {
                // get hour
                var hourStr = match.Groups[Constants.HourGroupName].Value;
                if (string.IsNullOrEmpty(hourStr))
                {
                    hourStr = match.Groups["hournum"].Value.ToLower();
                    if (!this.config.Numbers.TryGetValue(hourStr, out hour))
                    {
                        return ret;
                    }
                }
                else
                {
                    if (!int.TryParse(hourStr, out hour))
                    {
                        if (!this.config.Numbers.TryGetValue(hourStr.ToLower(), out hour))
                        {
                            return ret;
                        }
                    }
                }

                // get minute
                var minStr = match.Groups[Constants.MinuteGroupName].Value.ToLower();
                if (string.IsNullOrEmpty(minStr))
                {
                    minStr = match.Groups["minnum"].Value;
                    if (!string.IsNullOrEmpty(minStr))
                    {
                        min = this.config.Numbers[minStr];
                        hasMin = true;
                    }

                    var tensStr = match.Groups["tens"].Value;
                    if (!string.IsNullOrEmpty(tensStr))
                    {
                        min += this.config.Numbers[tensStr];
                        hasMin = true;
                    }
                }
                else
                {
                    min = int.Parse(minStr);
                    hasMin = true;
                }

                // get second
                var secStr = match.Groups[Constants.SecondGroupName].Value.ToLower();
                if (!string.IsNullOrEmpty(secStr))
                {
                    second = int.Parse(secStr);
                    hasSec = true;
                }
            }

            // Adjust by desc string
            var descStr = match.Groups[Constants.DescGroupName].Value.ToLower();

            // ampm is a special case in which at 6ampm = at 6
            if (config.UtilityConfiguration.AmDescRegex.Match(descStr).Success
                || config.UtilityConfiguration.AmPmDescRegex.Match(descStr).Success
                || match.Groups[Constants.ImplicitAmGroupName].Success)
            {
                if (hour >= Constants.HalfDayHourCount)
                {
                    hour -= Constants.HalfDayHourCount;
                }

                if (!config.UtilityConfiguration.AmPmDescRegex.Match(descStr).Success)
                {
                    hasAm = true;
                }
            }
            else if (config.UtilityConfiguration.PmDescRegex.Match(descStr).Success
                || match.Groups[Constants.ImplicitPmGroupName].Success)
            {
                if (hour < Constants.HalfDayHourCount)
                {
                    hour += Constants.HalfDayHourCount;
                }

                hasPm = true;
            }

            // adjust min by prefix
            var timePrefix = match.Groups[Constants.PrefixGroupName].Value.ToLower();
            if (!string.IsNullOrEmpty(timePrefix))
            {
                this.config.AdjustByPrefix(timePrefix, ref hour, ref min, ref hasMin);
            }

            // adjust hour by suffix
            var timeSuffix = match.Groups[Constants.SuffixGroupName].Value.ToLower();
            if (!string.IsNullOrEmpty(timeSuffix))
            {
                this.config.AdjustBySuffix(timeSuffix, ref hour, ref min, ref hasMin, ref hasAm, ref hasPm);
            }

            if (hour == 24)
            {
                hour = 0;
            }

            ret.Timex = "T" + hour.ToString("D2");
            if (hasMin)
            {
                ret.Timex += ":" + min.ToString("D2");
            }

            if (hasSec)
            {
                ret.Timex += ":" + second.ToString("D2");
            }

            if (hour <= Constants.HalfDayHourCount && !hasPm && !hasAm && !hasMid)
            {
                ret.Comment = Constants.Comment_AmPm;
            }

            ret.FutureValue = ret.PastValue = DateObject.MinValue.SafeCreateFromValue(year, month, day, hour, min, second);
            ret.Success = true;

            return ret;
        }
    }
}

File: Java/libraries/recognizers-text-date-time/src/main/java/com/microsoft/recognizers/text/datetime/parsers/BaseTimeParser.java
package com.microsoft.recognizers.text.datetime.parsers;

import com.google.common.collect.ImmutableMap;
import com.microsoft.recognizers.text.ExtractResult;
import com.microsoft.recognizers.text.ParseResult;
import com.microsoft.recognizers.text.datetime.Constants;
import com.microsoft.recognizers.text.datetime.TimeTypeConstants;
import com.microsoft.recognizers.text.datetime.parsers.config.ITimeParserConfiguration;
import com.microsoft.recognizers.text.datetime.parsers.config.PrefixAdjustResult;
import com.microsoft.recognizers.text.datetime.parsers.config.SuffixAdjustResult;
import com.microsoft.recognizers.text.datetime.utilities.ConditionalMatch;
import com.microsoft.recognizers.text.datetime.utilities.DateTimeFormatUtil;
import com.microsoft.recognizers.text.datetime.utilities.DateTimeResolutionResult;
import com.microsoft.recognizers.text.datetime.utilities.DateUtil;
import com.microsoft.recognizers.text.datetime.utilities.RegexExtension;
import com.microsoft.recognizers.text.datetime.utilities.TimeZoneResolutionResult;
import com.microsoft.recognizers.text.datetime.utilities.TimeZoneUtility;
import com.microsoft.recognizers.text.utilities.IntegerUtility;
import com.microsoft.recognizers.text.utilities.Match;
import com.microsoft.recognizers.text.utilities.RegExpUtility;
import com.microsoft.recognizers.text.utilities.StringUtility;

import java.time.LocalDateTime;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.regex.Pattern;

public class BaseTimeParser implements IDateTimeParser {

    private final ITimeParserConfiguration config;

    public BaseTimeParser(ITimeParserConfiguration config) {
        this.config = config;
    }

    @Override
    public String getParserName() {
        return Constants.SYS_DATETIME_TIME;
    }

    @Override
    public List<DateTimeParseResult> filterResults(String query, List<DateTimeParseResult> candidateResults) {
        return candidateResults;
    }

    @Override
    public ParseResult parse(ExtractResult extractResult) {
        return this.parse(extractResult, LocalDateTime.now());
    }

    @Override
    public DateTimeParseResult parse(ExtractResult er, LocalDateTime reference) {

        LocalDateTime referenceTime = reference;

        Object value = null;

        if (er.getType().equals(getParserName())) {
            DateTimeResolutionResult innerResult;

            // Resolve timezome
            if (TimeZoneUtility.shouldResolveTimeZone(er, config.getOptions())) {
                Map<String, Object> metadata = (Map)er.getData();
                ExtractResult timezoneEr = (ExtractResult)metadata.get(Constants.SYS_DATETIME_TIMEZONE);
                ParseResult timezonePr = config.getTimeZoneParser().parse(timezoneEr);

                innerResult = internalParse(er.getText().substring(0, er.getText().length() - timezoneEr.getLength()), referenceTime);

                if (timezonePr.getValue() != null) {
                    TimeZoneResolutionResult timeZoneResolution = ((DateTimeResolutionResult)timezonePr.getValue()).getTimeZoneResolution();
                    innerResult.setTimeZoneResolution(timeZoneResolution);
                }

            } else {
                innerResult = internalParse(er.getText(), referenceTime);
            }

            if (innerResult.getSuccess()) {
                ImmutableMap.Builder<String, String> futureResolution = ImmutableMap.builder();
                futureResolution.put(TimeTypeConstants.TIME, DateTimeFormatUtil.formatTime((LocalDateTime)innerResult.getFutureValue()));

                innerResult.setFutureResolution(futureResolution.build());

                ImmutableMap.Builder<String, String> pastResolution = ImmutableMap.builder();
                pastResolution.put(TimeTypeConstants.TIME, DateTimeFormatUtil.formatTime((LocalDateTime)innerResult.getPastValue()));

                innerResult.setPastResolution(pastResolution.build());

                value = innerResult;
            }
        }

        DateTimeParseResult ret = new DateTimeParseResult(
                er.getStart(),
                er.getLength(),
                er.getText(),
                er.getType(),
                er.getData(),
                value,
                "",
                value == null ? "" : ((DateTimeResolutionResult)value).getTimex());

        return ret;
    }

    protected DateTimeResolutionResult internalParse(String text, LocalDateTime referenceTime) {

        DateTimeResolutionResult innerResult = parseBasicRegexMatch(text, referenceTime);
        return innerResult;
    }

    // parse basic patterns in TimeRegexList
    private DateTimeResolutionResult parseBasicRegexMatch(String text, LocalDateTime referenceTime) {

        String trimmedText = text.trim().toLowerCase();
        int offset = 0;
        Optional<Match> match = Arrays.stream(RegExpUtility.getMatches(config.getAtRegex(), trimmedText)).findFirst();
        if (!match.isPresent()) {
            match = Arrays.stream(RegExpUtility.getMatches(config.getAtRegex(), config.getTimeTokenPrefix() + trimmedText)).findFirst();
            offset = config.getTimeTokenPrefix().length();
        }

        if (match.isPresent() && match.get().index == offset && match.get().length == trimmedText.length()) {
            return match2Time(match.get(), referenceTime);
        }

        // parse hour pattern, like "twenty one", "16"
        // create a extract comments which content the pass-in text
        Integer hour = null;
        if (config.getNumbers().containsKey(text)) {
            hour = config.getNumbers().get(text);
        } else {
            try {
                hour = Integer.parseInt(text);
            } catch (Exception ignored) {
                hour = null;
            }
        }

        if (hour != null && hour >= 0 && hour <= 24) {
            DateTimeResolutionResult result = new DateTimeResolutionResult();

            if (hour == 24) {
                hour = 0;
            }

            if (hour <= Constants.HalfDayHourCount && hour != 0) {
                result.setComment(Constants.Comment_AmPm);
            }

            result.setTimex(String.format("T%02d", hour));
            LocalDateTime value = DateUtil.safeCreateFromMinValue(referenceTime.getYear(), referenceTime.getMonthValue(), referenceTime.getDayOfMonth(), hour, 0, 0);
            result.setFutureValue(value);
            result.setPastValue(value);
            result.setSuccess(true);
            return result;
        }

        for (Pattern regex : config.getTimeRegexes()) {
            ConditionalMatch exactMatch = RegexExtension.matchExact(regex, trimmedText, true);

            if (exactMatch.getSuccess()) {
                return match2Time(exactMatch.getMatch().get(), referenceTime);
            }
        }

        return new DateTimeResolutionResult();
    }

    private DateTimeResolutionResult match2Time(Match match, LocalDateTime referenceTime) {

        DateTimeResolutionResult result = new DateTimeResolutionResult();
        boolean hasMin = false;
        boolean hasSec = false;
        boolean hasAm = false;
        boolean hasPm = false;
        boolean hasMid = false;
        int hour = 0;
        int minute = 0;
        int second = 0;
        int day = referenceTime.getDayOfMonth();
        int month = referenceTime.getMonthValue();
        int year = referenceTime.getYear();

        String writtenTimeStr = match.getGroup("writtentime").value;

        if (!StringUtility.isNullOrEmpty(writtenTimeStr)) {
            // get hour
            String hourStr = match.getGroup("hournum").value.toLowerCase();
            hour = config.getNumbers().get(hourStr);

            // get minute
            String minStr = match.getGroup("minnum").value.toLowerCase();
            String tensStr = match.getGroup("tens").value.toLowerCase();

            if (!StringUtility.isNullOrEmpty(minStr)) {
                minute = config.getNumbers().get(minStr);
                if (!StringUtility.isNullOrEmpty(tensStr)) {
                    minute += config.getNumbers().get(tensStr);
                }
                hasMin = true;
            }
        } else if (!StringUtility.isNullOrEmpty(match.getGroup("mid").value)) {
            hasMid = true;
            if (!StringUtility.isNullOrEmpty(match.getGroup("midnight").value)) {
                hour = 0;
                minute = 0;
                second = 0;
            } else if (!StringUtility.isNullOrEmpty(match.getGroup("midmorning").value)) {
                hour = 10;
                minute = 0;
                second = 0;
            } else if (!StringUtility.isNullOrEmpty(match.getGroup("midafternoon").value)) {
                hour = 14;
                minute = 0;
                second = 0;
            } else if (!StringUtility.isNullOrEmpty(match.getGroup("midday").value)) {
                hour = Constants.HalfDayHourCount;
                minute = 0;
                second = 0;
            }
        } else {
            // get hour
            String hourStr = match.getGroup(Constants.HourGroupName).value;
            if (StringUtility.isNullOrEmpty(hourStr)) {
                hourStr = match.getGroup("hournum").value.toLowerCase();
                if (!config.getNumbers().containsKey(hourStr)) {
                    return result;
                }

                hour = config.getNumbers().get(hourStr);
            } else {
                if (!IntegerUtility.canParse(hourStr)) {
                    if (!config.getNumbers().containsKey(hourStr.toLowerCase())) {
                        return result;
                    }

                    hour = config.getNumbers().get(hourStr.toLowerCase());
                } else {
                    hour = Integer.parseInt(hourStr);
                }
            }

            // get minute
            String minStr = match.getGroup(Constants.MinuteGroupName).value.toLowerCase();
            if (StringUtility.isNullOrEmpty(minStr)) {
                minStr = match.getGroup("minnum").value;
                if (!StringUtility.isNullOrEmpty(minStr)) {
                    minute = config.getNumbers().get(minStr);
                    hasMin = true;
                }

                String tensStr = match.getGroup("tens").value;
                if (!StringUtility.isNullOrEmpty(tensStr)) {
                    minute += config.getNumbers().get(tensStr);
                    hasMin = true;
                }
            } else {
                minute = Integer.parseInt(minStr);
                hasMin = true;
            }

            // get second
            String secStr = match.getGroup(Constants.SecondGroupName).value.toLowerCase();
            if (!StringUtility.isNullOrEmpty(secStr)) {
                second = Integer.parseInt(secStr);
                hasSec = true;
            }
        }

        // Adjust by desc string
        String descStr = match.getGroup(Constants.DescGroupName).value.toLowerCase();

        // ampm is a special case in which at 6ampm = at 6
        if (isAmDesc(descStr, match)) {
            if (hour >= Constants.HalfDayHourCount) {
                hour -= Constants.HalfDayHourCount;
            }

            if (!checkRegex(config.getUtilityConfiguration().getAmPmDescRegex(), descStr)) {
                hasAm = true;
            }

        } else if (isPmDesc(descStr, match)) {
            if (hour < Constants.HalfDayHourCount) {
                hour += Constants.HalfDayHourCount;
            }

            hasPm = true;
        }

        // adjust min by prefix
        String timePrefix = match.getGroup(Constants.PrefixGroupName).value.toLowerCase();
        if (!StringUtility.isNullOrEmpty(timePrefix)) {
            PrefixAdjustResult prefixResult = config.adjustByPrefix(timePrefix, hour, minute, hasMin);
            hour = prefixResult.hour;
            minute = prefixResult.minute;
            hasMin = prefixResult.hasMin;
        }

        // adjust hour by suffix
        String timeSuffix = match.getGroup(Constants.SuffixGroupName).value.toLowerCase();
        if (!StringUtility.isNullOrEmpty(timeSuffix)) {
            SuffixAdjustResult suffixResult = config.adjustBySuffix(timeSuffix, hour, minute, hasMin, hasAm, hasPm);
            hour = suffixResult.hour;
            minute = suffixResult.minute;
            hasMin = suffixResult.hasMin;
            hasAm = suffixResult.hasAm;
            hasPm = suffixResult.hasPm;
        }

        if (hour == 24) {
            hour = 0;
        }

        StringBuilder timex = new StringBuilder(String.format("T%02d", hour));

        if (hasMin) {
            timex.append(String.format(":%02d", minute));
        }

        if (hasSec) {
            timex.append(String.format(":%02d", second));
        }

        result.setTimex(timex.toString());

        if (hour <= Constants.HalfDayHourCount && !hasPm && !hasAm && !hasMid) {
            result.setComment(Constants.Comment_AmPm);
        }

        LocalDateTime resultTime = DateUtil.safeCreateFromMinValue(year, month, day, hour, minute, second);
        result.setFutureValue(resultTime);
        result.setPastValue(resultTime);

        result.setSuccess(true);

        return result;
    }

    private boolean isAmDesc(String descStr, Match match) {
        return checkRegex(config.getUtilityConfiguration().getAmDescRegex(), descStr) ||
                checkRegex(config.getUtilityConfiguration().getAmPmDescRegex(), descStr) ||
                !StringUtility.isNullOrEmpty(match.getGroup(Constants.ImplicitAmGroupName).value);
    }

    private boolean isPmDesc(String descStr, Match match) {
        return checkRegex(config.getUtilityConfiguration().getPmDescRegex(), descStr) ||
                !StringUtility.isNullOrEmpty(match.getGroup(Constants.ImplicitPmGroupName).value);
    }

    private boolean checkRegex(Pattern regex, String input) {
        Optional<Match> result = Arrays.stream(RegExpUtility.getMatches(regex, input)).findFirst();
        return result.isPresent();
    }
}
